### PR TITLE
✨ Allow Trunk configuration at a Port level

### PR DESCRIPTION
**Pull request by Xenwar**
Allow Trunk configuration at a Port level

This PR enables and disables port level trunk setting as follows.

- If user provides trunk field (true or false) at Port level, then that value is used for the port.
- If user does not provide trunk field at Port level, then the port inherits the trunk field of the instance. 

Example spec:
```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: OpenStackMachineTemplate
metadata:
  name: basic-1-control-plane
  namespace: default
spec:
  template:
    spec:
  template:
    spec:
      trunk: true
      ports:
      - description: "Port 1, no trunk specified"
      - description: "Port 2, trunk disabled"
        trunk: false
      - description: "Port 3, trunk enabled"
        trunk: true
      cloudName: openstack-1
      flavor: 2C-4GB-100GB
      identityRef:
        kind: Secret
        name: basic-1-cloud-config
      image: Ubuntu_20.04_node
```
Output:

```
(openstack) network trunk list  --sort-column Name
+--------------------------------------+-------------------------------+--------------------------------------+-----------------------------------------------------------+
| ID                                   | Name                          | Parent Port                          | Description                                               |
+--------------------------------------+-------------------------------+--------------------------------------+-----------------------------------------------------------+
| 03bbab0b-96d9-4226-bded-08ed8b3449ce | basic-1-control-plane-th8l2-0 | 6b6445f6-13af-4669-a301-cc79a79bf8ef | Created by cluster-api-provider-openstack cluster basic-1 |
| c09f979e-333f-4651-a4ff-93d7a72afc2d | basic-1-control-plane-th8l2-2 | fcd0a7ff-138c-4125-9f2a-42a2c32b13e5 | Created by cluster-api-provider-openstack cluster basic-1 |
+--------------------------------------+-------------------------------+--------------------------------------+-----------------------------------------------------------+
```

```
(openstack) port list -c Name -c Description -c ID --sort-column Name
+-------------------------------+----------------------------+--------------------------------------+
| Name                          | Description                | ID                                   |
+-------------------------------+----------------------------+--------------------------------------+
|                               |                            | 1dc33e63-3e4e-492a-86c4-70b304b58986 |
|                               |                            | 35830ee6-d984-4a1f-8e1a-703e024caca8 |
|                               |                            | e650d8ec-4786-4790-a6c9-befe7db57fae |
| basic-1-control-plane-th8l2-0 | Port 1, no trunk specified | 6b6445f6-13af-4669-a301-cc79a79bf8ef |
| basic-1-control-plane-th8l2-1 | Port 2, trunk disabled     | a944ec80-e147-41e8-aa6b-bbef1a47fe37 |
| basic-1-control-plane-th8l2-2 | Port 3, trunk enabled      | fcd0a7ff-138c-4125-9f2a-42a2c32b13e5 |
+-------------------------------+----------------------------+--------------------------------------+
(openstack) 
```

Signed-off-by: Anwar Hassen <anwar.hassen@est.tech>

### CODE REVIEW

**Comment by dulek** [comment]
This is awesome, I'm just interested if the logic here is fine. I'd expect that `trunk` setting at instance level would just change the default behavior for all ports but you could still be able to overwrite it by specifying it explicitly on the port level. If I understand the proposal correctly, currently the logic only allows that when instance level `trunk` is `true`.

**Comment by Xenwar** [comment]
> This is awesome, I'm just interested if the logic here is fine. I'd expect that `trunk` setting at instance level would just change the default behavior for all ports but you could still be able to overwrite it by specifying it explicitly on the port level. If I understand the proposal correctly, currently the logic only allows that when instance level `trunk` is `true`.

True. At the moment, instance level overrides all ports if it is set to false. 
If you have a use to make the port-level as a source of truth, we can make changes as well.

**Comment by Xenwar** [comment]
We can remove the conditional here https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/934/files#diff-5e7dbfd52305a21a11fb94fab201062ae2c7b851db93c2a8443a1d82d7aa7936R506
Then, ports would dictate trunk setting, regardless instance level value.

**Comment by dulek** [comment]
Thank you, this seems more logical to me. The main reason here is that I'm looking at this from Kuryr perspective which requires you to have trunk set only on the main port of the instance. This means that I can now set instance's `trunk` to false and only set `trunk` to true for that single port that requires it. The previous version of this code would force me to set `trunk` to false for each port that does not require it.

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [inline pkg/cloud/services/compute/instance.go:508]: This is a problem.

Previously if we specified trunk on the instance and a number of ports without any trunk setting, those ports would inherit the trunk setting of the instance. Here we're *always* using the trunk setting of the port, which now defaults to False rather than inheriting the trunk setting of the instance. This will cause a change of behaviour.

I think we want the following behaviour:
1. instance=true, port=undefined, trunk=true
2. instance=true, port=true, trunk=true
3. instance=true, port=false, trunk=false
4. instance=false, port=undefined, trunk=false
5. instance=false, port=true, trunk=true
6. instance=false, port=false, trunk=false

If we define port=undefined to be false (the behaviour here), that breaks case 1.
If we define port=undefined to be true, that breaks case 4.

IOW think we need to make portOpts.Trunk a *bool instead of a bool.

**Review by mdbooth — CHANGES_REQUESTED** [review]

**Comment by mdbooth** [comment]
Also, where are we on testing this kind of functionality?

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:197]: Tried to pass the instance.Tags as is. However, it is not always passed to the `createInstance`. line 195 is a guard against that.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:510]: Yes, this PR is creating a new trunk for every port marked as trunked. 
However, it is configurable as is shown in the description of this PR.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:508]: We are not inheriting instance level setting. As discussed with @dulek above, each port makes its own decision, regardless of what the instance says.
for (1) and (4), the check was on if a port is specified via `Portopts`, if not it is set to the instance level https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1673261708882c975a746bd0c2bdd5b62bca673a/pkg/cloud/services/compute/instance.go#L178

Is it so that there are other ways of specifying multiple ports ?

**Comment by Xenwar** [comment]
> Also, where are we on testing this kind of functionality?

I was waiting for the merge of https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/935 before adding tests, but add a test for it.

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:197]: I don't understand. What breaks?

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:508]: However, I think we previously *did* inherit the instance setting. Let's take a specific example:

```
      trunk: true
      ports:
      - description: "primary 1"
        foo: bar
      - description: "primary 2"
        foo: baz
```

Without this change, both of these ports will be trunked. This change will cause both of the above ports to now be trunked because `trunk` is not specified, and being a `bool` it will default to `false`. IOW this change causes a change in previously defined behaviour. I think that would be a regression.

I think the fix is simple: if we make `trunk` on the port a `*bool` instead of a `bool` we can define a third behaviour for the undefined case which is compatible with the previous behaviour.

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:510]: Please ignore this because it's out of scope for your change and consistent with the previous behaviour. However, now that I've seen it I don't understand why trunks are used this way. That could just be because I don't properly understand trunks, though.

**Review by mdbooth — CHANGES_REQUESTED** [review]
I still think this change would introduce a regression without changing `portopts.trunk` from `bool` to `*bool`.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:508]: Thanks. Now, I understood the consequence of not making it a pointer. 
Making changes to address it. Will also update the PR description accordingly.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:204]: Will try to get more specific data on this.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go]: Done.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:204]: Got some logs. 
Cause: When no tag is provided `OpenStackMachineTemplate.spec.template.spec.template.spec.tags`, the following error occurs. 

logs:
`E0716 20:27:52.384885     846 openstackmachine_controller.go:389] controller-runtime/manager/controller/openstackmachine "msg"="OpenStack instance cannot be created: error creating Openstack instance: Missing input for argument [Tags]" "error"="UpdateError" "cluster"="basic-1" "machine"="basic-1-control-plane-wsrg7" "name"="basic-1-control-plane-4p2jc" "namespace"="default" "openStackCluster"="basic-1" "openStackMachine"="basic-1-control-plane-4p2jc" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" `

The problem does not occur for all ports when user specified multiple ports. It only occurs to some of them. Which is the reason why I added that check.

**Comment by Xenwar** [comment]
@mdbooth  added an e2e test

File: api/v1alpha4/types.go
Comment (adduarte) [inline api/v1alpha4/types.go:133]: just a nit. is this comment correct(needed?)  wouldn't "network level field" be openStackMachine.Spec.network.Trunk? 
But the port inherits openStackMachine.Spec.Trunk (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/934/files#diff-5e7dbfd52305a21a11fb94fab201062ae2c7b851db93c2a8443a1d82d7aa7936R146)

Perhaps drop the comment to avoid confusion..

File: pkg/cloud/services/compute/instance.go
Comment (adduarte) [inline pkg/cloud/services/compute/instance.go:156]: "inherits  openStackMachine.Spec.Trunk"

File: api/v1alpha4/types.go
Comment (Xenwar) [reply api/v1alpha4/types.go:133]: True, will correct it. We need the comment is it is needed, visible when running `kubectl explain`
I am also working on PortSecurity, confused the two.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go]: done

File: pkg/cloud/services/compute/instance.go
Comment (dulek) [reply pkg/cloud/services/compute/instance.go:516]: @mdbooth: It's a Neutron trunk, which might be a bit confusing. In Neutron a *port* can have a trunk created and then you add other ports as *subports* in that trunk. In Kuryr we use that to dynamically add subports which serve as pods interfaces. What you're proposing is doable, but I'm not sure if I see value in subports added statically. Anyway if we'd want to implement that, I'd rather add a `subports` field in `PortOpts` and get subports defined there.

File: pkg/cloud/services/compute/instance.go
Comment (dulek) [reply pkg/cloud/services/compute/instance.go]: Thanks @mdbooth, I missed this detail.

File: pkg/cloud/services/compute/instance.go
Comment (adduarte) [reply pkg/cloud/services/compute/instance.go:516]: @dulek is the name of the trunk singinificant to kuryr? Does kuryr look up trunks per port (parent port.name for example), or by instance

**Comment by Xenwar** [comment]
Once trunk is supported, a test for trunk with the following content can be added https://gist.github.com/Xenwar/d2a9a120531aef09f7670b57d43e0857

**Comment by Xenwar** [comment]
The trunk feature is tested on separate openstack environment with trunk support.
I was unable to add the E2E test due to the CAPO test infra does not support trunk yet, report https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/961

Once it supports trunk, the following test can be added, PortSecurity is also tested
https://gist.github.com/Xenwar/d2a9a120531aef09f7670b57d43e0857

File: pkg/cloud/services/compute/instance.go
Comment (tobiasgiese) [inline pkg/cloud/services/compute/instance.go:159]: You have to create the variable `pOpts` to not override `port.Trunk`, correct?
Or can we do sth like this? And pass `PortOpts: &port` in l.164 and l.172

```suggestion
		// No Trunk field specified for the port, inherits openStackMachine.Spec.Trunk.
		if port.Trunk == nil {
			port.Trunk = &openStackMachine.Spec.Trunk
		}
```

File: pkg/cloud/services/compute/instance.go
Comment (tobiasgiese) [inline pkg/cloud/services/compute/instance.go:546]: I would keep consistent to the other methods and keep the `clusterName` in the arguments. We're doing this several times like this

**Review by tobiasgiese — COMMENTED** [review]
two nits

**Comment by jichenjc** [comment]
> 
> 
> The trunk feature is tested on separate openstack environment with trunk support.
> I was unable to add the E2E test due to the CAPO test infra does not support trunk yet, report #961
> 
> Once it supports trunk, the following test can be added, PortSecurity is also tested
> https://gist.github.com/Xenwar/d2a9a120531aef09f7670b57d43e0857

thanks for the info, let's merge this after review comments done,  I agree we can track the issue and fix later on

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go]: Thanks. done.

File: pkg/cloud/services/compute/instance.go
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:152]: pOpts is created to avoid code duplication. 
The check needs to be done for each Port over the loop.

**Comment by Xenwar** [comment]
I have addressed on the of the comments. @tobiasgiese would you please check again to see if I have addresses the issues. 
Thanks.

File: pkg/cloud/services/compute/instance.go
Comment (tobiasgiese) [reply pkg/cloud/services/compute/instance.go]: But you don't have to set `pOpts := &openStackMachine.Spec.Ports[i]` if `port` is already `openStackMachine.Spec.Ports[i]` because of the for loop

**Comment by Xenwar** [comment]
@jichenjc
The last comments are addressed. Do we merge it now ?
Thanks.

**Comment by jichenjc** [comment]
/approve 

thanks a lot for your help~

**Pull request merged**
